### DISCUSSION TITLE
Can you generate by some steps and save the intermediate result , then generate the other steps?

### DISCUSSION THREAD

**Discussion by Slarper** [comfyanonymous/ComfyUI#220] [General]
![image](https://user-images.githubusercontent.com/98243861/226971280-d63fe627-8c86-4a42-903b-0b8412bde419.png)

I used  two 14-step KSamplers in series to generate the left one, and used a 28-step KSampler for the right one. But they are obviously different. But why? I think `14 steps + 14 steps = 28 steps` so they should be the same result.

My workflow:

[ComfyUI_00045_.zip](https://github.com/comfyanonymous/ComfyUI/files/11042391/ComfyUI_00045_.zip)

(The model named after `nai` is the leak novelai model)

Imagination result: 

![image](https://user-images.githubusercontent.com/98243861/226978604-6cee139c-1901-471d-bdbf-7dd2074fd1c1.png)

**Comment by Gasskin** [comment]
i think these two ways is totally different...

in my opinion,the second KSample will not continue the work from the first one...because that the KSample has no pause/continue state

every time start a KSample,it will start from step0,buy in your case, the second one's input params is the first one's output,and the first one's input is text description,for that,the finally result is just the result of KSample2，not KSample2 continue from KSample1

so step14+step14 will no same as step28..

but i am also new here,the above is just my personal understanding,maybe its wrong

**Comment by Slarper** [comment]
After testing, I get it.

![image](https://user-images.githubusercontent.com/98243861/227150284-c27b0b78-1a19-4680-9e38-4312bb4fffa6.png)

Let me explain it for you.

**First, the 14-step KSampler and the 28-step KSampler are totally different from each other.** Thanks to [Raveness](https://discord.com/channels/752483298198028309/1084878026644393984/1088310725920960632):

> Think of it more like 0-14 of 14 is doing the whole task in 14, while 0-14 of 28 is doing half of the task in smaller steps

The 14-step KSampler divides the work into 14 pieces while the 28-step KSampler divides the work into 28 smaller pieces. So their works can not replace each other.

So you need two 28-step KSamplers, each one does half of the work.

the fomula `14 steps + 14 steps = 28 steps` should be modified into
`0-14 steps of 28-step KS + 14-28 steps of 28-step KS = 0-28 steps of 28-step KS`. 
Notice that even each step of the same KSampler are different:
 `0-14 steps of 28-step KS + 0-14 steps of 28-step KS = 0-28 steps of 28-step KS` is wrong. (below)

![image](https://user-images.githubusercontent.com/98243861/227157572-91b08fe8-a7d4-4c4f-82c7-d5126db26401.png)

In short:

![image](https://user-images.githubusercontent.com/98243861/227155847-02d66699-4cbd-4fe5-8beb-fdb2db637bf0.png)

**Second, you should be careful with the add/clear noise.** 

For the two KS in series, the first KS should enable `return_with_leftover_noise` while the second KS should disable `add_noise`.

![image](https://user-images.githubusercontent.com/98243861/227159255-232c159e-e455-4df0-8ab9-df8c8316a74d.png)

Think of it like this workflow:

```
add noise
iterating 1
iterating 2
...
iterating 28
clear leftover noise
```

if you do not enable `return_with_leftover_noise` of the first one and disable `add_noise` of the second one, the workflow will be like:

```
add noise
iterating 1
iterating 2
...
iterating 14
+ clear leftover noise
+ add noise
iterating 15
...
iterating 28
clear leftover noise
```

that will cause the result a bit of different.

For more KSs in series, make sure only the first one adds noise and only the last one clears leftover noise.

Here is the workflow.

[ComfyUI_00061_.zip](https://github.com/comfyanonymous/ComfyUI/files/11049181/ComfyUI_00061_.zip)

**Reply by Gasskin** [reply]
i got it! thanks, learn a lot.
but what can we do with this? its a bit strange..
maybe we can use checkpoint1 int step0-14, and use checkpoint2 in step14-28?

**Reply by Slarper** [reply]
I study it for outputting the intermediate steps as images. (the below one is only 14/28 steps)

![image](https://user-images.githubusercontent.com/98243861/227164099-97ae21c6-6c39-4951-b1ad-b38ce1463f87.png)

**Reply by Slarper** [reply]
> maybe we can use checkpoint1 int step0-14, and use checkpoint2 in step14-28?

Yes.

**Reply by Gasskin** [reply]
cool!

**Comment by Slarper** [comment]
BTW, according to our conclusion, this picture is WRONG.

![image](https://user-images.githubusercontent.com/98243861/227165275-2eac75bb-5322-4aa2-9665-8d11677e86bb.png)

source : https://imgur.com/a/VjFi5uM

It's NOT the way how the AI draw.

Because the way to generate this X/Y/Z plot by au111 is like:

> Use a 3-step KS to generate the first image, then use a 5-step KS to generate the second image, then use a 7-step KS, then 10-step KS.

It means that to generate the whole image needs 3+5+7+10 = 25 iterations totally.

But according to our conlusion1, n-step KSs cannot replace each other when n is varied. 
So it just generates a 3/3 image, a 5/5 image, a 7/7 image and a 10/10 image. 
But it should be a 3/10 image, a 5/10 image, a 7/10 image and a 10/10 image.

The real way how AI draw is like (in ComfyUI):

![image](https://user-images.githubusercontent.com/98243861/227169835-50d94b7c-476a-4b81-b52c-c5eec866db71.png)

And it only takes 10 iterations! (vs. 3+5+7+10 = 25 iterations in au111)

here is the workflow, you can test it yourself.

[ComfyUI_00071_.zip](https://github.com/comfyanonymous/ComfyUI/files/11049490/ComfyUI_00071_.zip)

Another reason to turn to ComfyUI.☺️

**Reply by Slarper** [reply]
@Gasskin

**Reply by Gasskin** [reply]
i think the 3/5/7/10 means different stages of the same KS...
3 means step3, 5 means step5 and so on...

**Reply by Slarper** [reply]
> i think the 3/5/7/10 means different stages of the same KS... 3 means step3, 5 means step5 and so on...

Yes but the original is wrong, it actually generates images by different KSs. My version is correct.
